### ISSUE TITLE
Is it possible to open the terminal without moving NvimTree?

### ISSUE THREAD

**Issue by davidlunadeleon** [akinsho/toggleterm.nvim#97]
Steps to reproduce:

1. Open file and open NvimTree.
2. Toggle terminal to open it.

Current behavior:

1. The terminal occupies the whole bottom of the screen, moving the window for NvimTree up.
2. Toggling Nvim-Tree to close and open again results in the expected layout. The tree to the side, and both terminal and file next to it.
3. If the terminal is closed and opened again, the layout returns to NvimTree not being correctly displayed again.

Expected behavior:

- Opening the terminal, doesn't affect the positioning of NvimTree if it's open.
- If NvimTree is open, the terminal doesn't take the whole bottom of the screen.
- If NvimTree is closed, the terminal can take the whole bottom of the screen.

My question is. Is this something that can already be configured? Sorry if it's possible and I missed it 😅

**Comment by akinsho** [comment]
Hi @davidlunadeleon,

Not really, this is a lot more complicated than it seems for a few different reasons,
generally sidebars are just splits like any other, although it's an increasing trend in nvim to have sidebars and things, nvim doesn't treat these windows any differently from other windows, so any behaviour for splits affects all splits. Technically, we could read the window layout as a nested list and try and see if the terminal is beside a particular buffer but having done that in another plugin, I'm 100% against the complexity that brings.

Currently, the way this plugin splits windows is that it's designed to split under all existing windows (or if vertical beside all horizontal), otherwise if you had 2 vertical splits and toggled the terminal it would appear under just one of the vertical splits which is not acceptable IMO and because of the earlier point not solvable (easily/desirably).

In short, there isn't an easy solution to the problem and the only potential solution introduces a large amount of complexity which I think far outweighs the benefits. This sort of problem would be easier to deal with if nvim special cased sidebar splits, so they weren't affected by `wincmd` in the same way, but that's super unlikely to get merged upstream.

**Comment by davidlunadeleon** [comment]
Thank you for the detailed answer! 🙂

It really sounds more complicated than I initially thought it would be, and definitely not something that would be solved in here. Guess that the people migrating from VSCode and other editors are used to the sidebar being there.

In case anyone wants something similar in the future, I'll leave my hacky solution in here:
```
vim.api.nvim_set_keymap('n', '<C-n>', '<cmd>:NvimTreeToggle<cr>', {})
vim.api.nvim_set_keymap('n', '<leader>to', [[<cmd>:ToggleTermOpenAll<cr><C-\><C-n><C-n><C-n>]], { silent = true })
```
Some notes:
- It causes a quick flicker since the tree is closed and opened, but it results in the desired window positioning.
- The selected window will end up being NvimTree.

**Comment by Enside** [comment]
Recently I find a way solve the problem perfectly, set this function to `on_open` of toggleterm's config
```lua
function set_nvimtree_when_open_term(terminal)
  local nvimtree = require "nvim-tree"
  local nvimtree_view = require "nvim-tree.view"
  if nvimtree_view.is_visible() and terminal.direction == "horizontal" then
    local nvimtree_width = vim.fn.winwidth(nvimtree_view.get_winnr())
    nvimtree.toggle()
    nvimtree_view.View.width = nvimtree_width
    nvimtree.toggle(false, true)
  end
end
```

**Comment by denstiny** [comment]
```lua
M.set_nvimtree_when_open_term = function()
  local nvimtree_view = require "nvim-tree.view"
  if nvimtree_view.is_visible() then
    vim.cmd("NvimTreeToggle")
    vim.cmd("ToggleTerm")
    require('nvim-tree').toggle(false, true)
  else
    vim.cmd("ToggleTerm")
  end
end
```

**Comment by askfiy** [comment]
This is an expected feature ..

**Comment by akinsho** [comment]
> This is an expected feature ..

Is a pretty incorrect statement for *a lot of reasons*, I'd rather not go into explaining all the details of how windows work in neovim, but a few points.

1. The sidebar panel of *any plugin* is just a window, neovim does not discriminate, just because you put a file tree in a window doesn't make that window magic, it is still expected to move and be moved just as any other buffer in a window would do.
2. Neovim does not provide any magic tools so that a plugin author can just automatically make sure that nvim tree or some other sidebar panel never affects toggleterm, toggleterm is just another window the same as nvim tree
3. An integration with nvim tree or n number of file tree plugins that use a bunch of different layouts and filetypes is not "expected" since that puts the burden on this plugin of supporting n number of "popular" file tree or outline plugins
4. This project is FOSS and so no features are "expected"

It might not have been intended, but comments like this are not helpful. At the very least if *you* think it is expected then since this is FOSS you should get *your* hands dirty and contribute.

**Comment by akinsho** [comment]
Or better still ask the other of nvim tree to add functionality to that plugin so that if it is not full height at any point then it re-arranges itself 🤷🏿‍♂️.

**Comment by askfiy** [comment]
I see what you mean, sorry it might be a translation issue, I mean it makes sense, but should let people implement it themselves, because there are many plugins for the sidebar, like undotree, nvim-tree, neotree, etc., these It needs to be maintained by the user..

sorry for misunderstanding

**Comment by askfiy** [comment]
I researched for some time and now it should support 3 types of sidebars `nvim-tree`, `undotree`, `dbui`.

If you are interested, you can try:

```
function aux.get_all_win_buf_ft()
    local result = {}
    local wins = vim.api.nvim_list_wins()

    for _, win_id in ipairs(wins) do
        if vim.api.nvim_win_is_valid(win_id) then
            local buf_id = vim.api.nvim_win_get_buf(win_id)
            table.insert(result, {
                win_id = win_id,
                buf_id = buf_id,
                buf_ft = vim.api.nvim_buf_get_option(buf_id, "filetype"),
            })
        end
    end

    return result
end


function aux.terminal_offset_run_command(command)
    local offset_ft = {

        NvimTree = function(win_id)
            vim.cmd("NvimTreeToggle")
            vim.cmd(command)
            require("nvim-tree").toggle(false, true)
        end,

        undotree = function(win_id)
            vim.g.undotree_SetFocusWhenToggle = 0
            vim.cmd("UndotreeToggle")
            vim.cmd(command)
            vim.cmd("UndotreeToggle")
            vim.g.undotree_SetFocusWhenToggle = 1
        end,

        dbui = function(win_id)
            vim.api.nvim_win_close(win_id, true)
            vim.cmd(command)
            vim.cmd("DBUIToggle")

            local max_term_id = 0
            local max_win_id = 0

            for _, opts in ipairs(aux.get_all_win_buf_ft()) do
                if opts.buf_ft == "toggleterm" then
                    local buf_name = vim.api.nvim_buf_get_name(opts.buf_id)
                    local term_id = tonumber(buf_name:match("#(%d+)$"))
                    max_term_id = math.max(max_term_id, term_id)
                    max_win_id = opts.win_id
                end
            end

            if max_win_id ~= 0 then
                vim.fn.win_gotoid(max_win_id)
            end
        end,
    }

    for _, opts in ipairs(aux.get_all_win_buf_ft()) do
        if vim.tbl_contains(vim.tbl_keys(offset_ft), opts.buf_ft) then
            offset_ft[opts.buf_ft](opts.win_id)
            return
        end
    end

    vim.cmd(command)
end
```

use it:

```
local count = vim.api.nvim_eval("v:count1")
aux.terminal_offset_run_command(string.format("exe %d.'ToggleTerm'", count))

==========

aux.terminal_offset_run_command("ToggleTermToggleAll")
```

Notice!!! When opening multiple terminals, if you open `dbui` then every time you redisplay the terminal it will always focus on the last terminal.

**Comment by xiantang** [comment]
for nerdtree

```lua
require("toggleterm").setup {
  start_in_insert = false,
  size = 20,
  on_open = function(_)
    -- check nerdtree is visible
    local name = vim.fn.bufname("NERD")
    local winnr = vim.fn.bufwinnr(name)
    if winnr ~= -1 then
      local pwd = vim.fn.getcwd()
      -- NerdTree
      -- async running
      vim.defer_fn(function()
        local cmd = string.format("NERDTreeToggle %s ", pwd)
        vim.cmd(cmd)
        local cmd_retoggle = string.format("NERDTreeToggle %s | 1 | wincmd p", pwd)
        vim.cmd(cmd_retoggle)
      end, 100)


    end
  end,
}
```
